lib/rule-tester/rule-tester.js

```diff
@@ -304,6 +305,41 @@ function throwForbiddenMethodError(methodName, prototype) {
     };
 }
 
+/**
+ * Extracts names of {{ placeholders }} from the reported message.
+ * @param   {string} message Reported message
+ * @returns {string[]} Array of placeholder names
+ */
+function getMessagePlaceholders(message) {
+    const matcher = getPlaceholderMatcher();
+
+    return Array.from(message.matchAll(matcher), ([, name]) => name.trim());
+}
+
+/**
+ * Returns the placeholders in the reported messages but
+ * only includes the placeholders available in the raw message if data is not specified or
+ * excludes introduced placeholders specified in the data values.
+ * @param {string} message The reported message
+ * @param {string} raw The raw message specified in the rule meta.messages
+ * @param {undefined|Record<unknown, unknown>} data The passed
+ * @returns {string[]} Missing placeholder names
+ */
+function getUnsubstitutedMessagePlaceholders(message, raw, data = {}) {
+    const unsubstituted = getMessagePlaceholders(message);
+
+    if (unsubstituted.length === 0) {
+        return [];
+    }
+
+    // Remove false positives by only counting placeholders in the raw message, which were not provided in the data matcher or added with a data property
+    const known = getMessagePlaceholders(raw);
+    const provided = Object.keys(data);
+    const added = new Set(Object.values(data).flatMap(value => (typeof value === "string" ? getMessagePlaceholders(value) : [])));
+
+    return unsubstituted.filter(name => known.includes(name) && !provided.includes(name) && !added.has(name));
```

Comment:
I'm not sure if we need `&& !added.has(name)` here. Is there a case where it removes more false positives? I can only think of cases where it might introduce false negatives (though, those are extreme edge cases).

Reply:
For cases like in [no-restricted-syntax](https://github.com/eslint/eslint/blob/15c143f96ef164943fd3d39b5ad79d9a4a40de8f/tests/lib/rules/no-restricted-syntax.js#L123) where a case "adding" a new placeholder  by using placeholders via a data value (not the same name as the data property)

Reply:
> For cases like in [no-restricted-syntax](https://github.com/eslint/eslint/blob/15c143f96ef164943fd3d39b5ad79d9a4a40de8f/tests/lib/rules/no-restricted-syntax.js#L123) where a case "adding" a new placeholder by using placeholders via a data value (not the same name as the data property)

Wouldn't that be already filtered out by `known.includes(name) && !provided.includes(name)`?